File: IDAUtil/Model/Properties/TcodeProperty/VA02Obj/OrderObjects/QtyConversions/QtyConversionsOrderPropertyFactory.vb
Public Module QtyConversionsOrderPropertyFactory
    Public Function createSalesDocumentList(zv04IList As List(Of ZV04IProperty)) As List(Of QtyConversionOrderProperty)
        Dim baseSalesDocumentQuery =
        From q In zv04IList
        Group q By line = New With {Key q.order, q.shipTo, q.soldTo, q.shipToName, q.docDate, q.pONumber, q.delivery}
        Into Group
        Select base = New QtyConversionOrderProperty With {
            .orderNumber = line.order,
            .deliveryNumber = line.delivery,
            .docDate = line.docDate,
            .pONumber = line.pONumber,
            .shipTo = line.shipTo,
            .soldTo = line.soldTo,
            .soldToName = line.shipToName,
            .documentLineList = Group.Aggregate(New List(Of DocumentLine), Function(x, q)
                                                                               x.Add(New DocumentLine() With {.item = q.item, .material = q.material, .quantity = q.orderQty})
                                                                               Return x
                                                                           End Function),
            .documentLineChangeList = .documentLineList
        }
        Return baseSalesDocumentQuery.ToList()
    End Function
End Module

File: IDAUtil/Model/Properties/TcodeProperty/VA02Obj/OrderObjects/QtyConversions/QtyConversionsOrderPropertyFactory.cs
using IDAUtil.Model.Properties.TcodeProperty.ZV04Obj;
using System.Collections.Generic;
using System.Linq;

namespace IDAUtil {
    public static class QtyConversionsOrderPropertyFactory {
        public static List<QtyConversionOrderProperty> createSalesDocumentList(List<ZV04IProperty> zv04IList) {

            var baseSalesDocumentQuery = (from q in zv04IList
                                          group q by new { q.order, q.shipTo, q.soldTo, q.shipToName, q.docDate, q.pONumber, q.delivery } into line
                                          select new QtyConversionOrderProperty() {
                                              orderNumber = line.Key.order,
                                              deliveryNumber = line.Key.delivery,
                                              docDate = line.Key.docDate,
                                              pONumber = line.Key.pONumber,
                                              shipTo = line.Key.shipTo,
                                              soldTo = line.Key.soldTo,
                                              soldToName = line.Key.shipToName,
                                              documentLineList = line.Aggregate(new List<DocumentLine>(), (x, q) => {
                                                  x.Add(new DocumentLine() { item = q.item, material = q.material, quantity = q.orderQty });
                                                  return x;
                                              }),
                                              documentLineChangeList = line.Aggregate(new List<DocumentLine>(), (x, q) => {
                                                  x.Add(new DocumentLine() { item = q.item, material = q.material, quantity = q.orderQty });
                                                  return x;
                                              })
                                          });

            return baseSalesDocumentQuery.ToList();
        }
    }
}